pkg/connector/login.go

```diff
@@ -0,0 +1,311 @@
+			ID:          "qr",
+		},
+		{
+			Name:        "Pairing code",
+			Description: "Input your phone number to get a pairing code, to pair the bridge to your WhatsApp account",
+			ID:          "pairing-code",
+		},
+	}
+}
+
+func (wa *WhatsAppConnector) CreateLogin(_ context.Context, user *bridgev2.User, flowID string) (bridgev2.LoginProcess, error) {
+
+	if flowID == "qr" {
+		return &QRLogin{User: user, Main: wa}, nil
+	} else if flowID == "pairing-code" {
+		return &PairingCodeLogin{User: user, Main: wa}, nil
+	} else {
+		return nil, fmt.Errorf("invalid login flow ID")
+	}
+}
+
+func makeQRChan(connector *WhatsAppConnector, user *bridgev2.User) (client *whatsmeow.Client, qrChan <-chan whatsmeow.QRChannelItem, cancelFunc context.CancelFunc, err error) {
+	log := connector.Bridge.Log.With().
+		Str("action", "login").
+		Stringer("user_id", user.MXID).
+		Logger()
+	qrCtx, cancel := context.WithCancel(log.WithContext(context.Background()))
+	cancelFunc = cancel
+
+	device := connector.DeviceStore.NewDevice()
+
+	wa := &WhatsAppClient{
+		Main:   connector,
+		Device: device,
+	}
+
+	wa.MakeNewClient(log)
+
+	client = wa.Client
+
+	qrChan, err = client.GetQRChannel(qrCtx)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	err = client.Connect()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return
+}
+
+func makeUserLogin(ctx context.Context, user *bridgev2.User, client *whatsmeow.Client) (ul *bridgev2.UserLogin, err error) {
+	newLoginID := waid.MakeWAUserLoginID(client.Store.ID)
+	ul, err = user.NewLogin(ctx, &database.UserLogin{
+		ID:         newLoginID,
+		RemoteName: client.Store.PushName,
+		Metadata: &UserLoginMetadata{
+			WADeviceID: client.Store.ID.Device,
+		},
+	}, &bridgev2.NewLoginParams{
+		DeleteOnConflict: true,
+	})
+	return
+}
+
+type QRLogin struct {
+	User       *bridgev2.User
+	Main       *WhatsAppConnector
+	Client     *whatsmeow.Client
+	cancelChan context.CancelFunc
+	QRChan     <-chan whatsmeow.QRChannelItem
+
+	AccData *store.Device
+}
+
+var _ bridgev2.LoginProcessDisplayAndWait = (*QRLogin)(nil)
+
+func (qr *QRLogin) Cancel() {
+	qr.cancelChan()
+	go func() {
+		for range qr.QRChan {
+		}
+	}()
+}
+
+const (
+	LoginStepQR               = "fi.mau.whatsapp.login.qr"
+	LoginStepComplete         = "fi.mau.whatsapp.login.complete"
+	LoginStepPairingCodeInput = "fi.mau.whatsapp.login.pairing.code"
+	LoginStepPairingCode      = "fi.mau.whatsapp.login.pairing.code"
+)
+
+func (qr *QRLogin) Start(ctx context.Context) (*bridgev2.LoginStep, error) {
+	client, qrChan, cancelFunc, err := makeQRChan(qr.Main, qr.User)
+	if err != nil {
+		return nil, err
+	}
+	qr.QRChan = qrChan
+	qr.cancelChan = cancelFunc
+	qr.Client = client
+	var resp whatsmeow.QRChannelItem
+
+	select {
+	case resp = <-qr.QRChan:
+		if resp.Error != nil {
+			return nil, resp.Error
+		} else if resp.Event != "code" {
+			return nil, fmt.Errorf("invalid QR channel event: %s", resp.Event)
+		}
+	case <-ctx.Done():
+		qr.cancelChan()
```

Comment:
maybe should call `qr.Cancel` directly here so that we have a single place to do cleanup of the channel?